rpcs3/Emu/RSX/Program/ProgramStateCache.h

```diff
@@ -202,31 +225,52 @@ class program_state_cache
 	}
 
 	/// bool here to inform that the program was preexisting.
-	std::tuple<const fragment_program_type&, bool> search_fragment_program(const RSXFragmentProgram& rsx_fp, bool force_load = true)
+	std::tuple<const fragment_program_type&, bool> search_fragment_program(const RSXFragmentProgram& rsx_fp, usz /*rsx_fp_invalidation_count*/)
 	{
 		bool recompile = false;
 		typename binary_to_fragment_program::iterator it;
 		fragment_program_type* new_shader;
-
 		{
+			thread_local const std::pair<const RSXFragmentProgram, fragment_program_type>* prev_fp = nullptr;
```

Comment:
Isn't this pointer invalidated if the container has an insert done on it?
This whole approach is q bit too convoluted. Why integer counter instead of bool? Why thread-local instead of atomic? etc.
I get the general idea, but it requires a much more involved implementation to be "proper".

Reply:
Yep, unordered_map does. I set fetched invalidation count to umax in this case, I can also set to pointer to be null for the intention to be cleaer.